numpy/core/src/multiarray/dragon4.c

```diff
@@ -1028,18 +1087,20 @@ Dragon4(const npy_uint64 mantissa, const npy_int32 exponent,
      * scaledMarginHigh is a pointer so that it can point to scaledMarginLow in
      * the case they must be equal to each other, otherwise it will point to
      * optionalMarginHigh.
+     *
+     * These are static to save stack space, so this function is not re-entrant.
      */
-    BigInt scale;
-    BigInt scaledValue;
-    BigInt scaledMarginLow;
+    static BigInt scale;
+    static BigInt scaledValue;
+    static BigInt scaledMarginLow;
     BigInt *scaledMarginHigh;
-    BigInt optionalMarginHigh;
+    static BigInt optionalMarginHigh;
 
     const npy_float64 log10_2 = 0.30102999566398119521373889472449;
```

Comment:
Does this need more digits for IBM double double?

Reply:
no, the exact value here isn't super important, it just optimizes some branch prediction.

Reply:
This value is Discussed on Ryan Juckett's website under "logarithms", plus in the Burger & Dybvig paper.

http://www.ryanjuckett.com/programming/printing-floating-point-numbers/

Reply:
Presumably having more digits would further optimize the prediction though? Or is this already more digits than an IBM double double can hold?

Reply:
Well, in some sense the main innovation of Dragon4 and the Burger & Dybvig paper is the fact that this calculation uses a simple float64 multiplication! And the calculation is already inexact enough that these extra digits shouldn't matter.

This value of `log10_2` is used in only one place, in the calculation of `digitExponent`. Note there is a long comment there (not by me) partly explaining this calculation, and you can also read the intro to the Burger & Dybvig which explains the motivation, and section 3.2 is entirely dedicated to exlaining it. The goal is to calculate a the base-10 exponent. Older algorithms (Steel and White) used a very expensive BigInt calculation to do this, which dominated the cost of the algorithm. Burger & Dybvig's contribution was to realize that if you just use a simple float64 calculation, you would at most be off by 1, and you could later detect this and correct it. That's why there's two branches: A "correct by one" path and a "no correction" path. Ryan Jucket realized that by adding an additional constant you could get it to go into the faster path more often, without risk of being off by more than one.

Just to try to illustrate, here is the actual calculation:
```
digitExponent = ceil( (mantissa - exponent)*log10_2 + 0.69)
```
Ryan Juckett added that extra 0.69, noting that it "moves the error range from about (-0.301, 0.0] to about (-0.991, -0.69]". We want to move the lower bound as close to -1 as possible, so in principle we could even use 0.699. But there is some danger getting too close to -1 since there is error in the float64 multiply with `log10_2` itself, B&D estimate it is up to `10**-10` for float64 mantissas and I estimate it is `10**-8` for float80 mantissas.

In summary, the computation is already off by about 0.009, meaning branch prediciton could be optimized by this amount, but that's so small it doesn't seem worth touching.

Reply:
My misunderstanding here was not noticing that this was a float64 instead of a longdouble - but thanks for the explanation anyway!